homeassistant/components/hunterdouglas_powerview/__init__.py

```diff
@@ -1 +1,196 @@
+    schema = vol.Schema(vol.Unique())
+    schema(hosts)
+    return value
+
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.All(cv.ensure_list, [DEVICE_SCHEMA], _has_all_unique_hosts)},
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+PLATFORMS = ["cover", "scene"]
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, hass_config: dict):
+    """Set up the Hunter Douglas PowerView component."""
+    hass.data.setdefault(DOMAIN, {})
+
+    if DOMAIN not in hass_config:
+        return True
+
+    for conf in hass_config[DOMAIN]:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
+            )
+        )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Hunter Douglas PowerView from a config entry."""
+
+    config = entry.data
+
+    hub_address = config.get(CONF_HOST)
+    websession = async_get_clientsession(hass)
+
+    pv_request = AioRequest(hub_address, loop=hass.loop, websession=websession)
+
+    hub = Hub(pv_request)
+
+    try:
+        async with async_timeout.timeout(10):
+            await hub.query_user_data()
+    except HUB_EXCEPTIONS:
+        _LOGGER.error("Connection error to PowerView hub: %s", hub_address)
+        raise ConfigEntryNotReady
+    if not hub.ip:
+        _LOGGER.error("Unable to initialize PowerView hub: %s", hub_address)
+        raise ConfigEntryNotReady
+
+    device_info = await _async_get_device_info(hub, pv_request)
+
+    rooms = Rooms(pv_request)
+    room_data = _async_map_data_by_id((await rooms.get_resources())[ROOM_DATA])
+
+    scenes = Scenes(pv_request)
+    scene_data = _async_map_data_by_id((await scenes.get_resources())[SCENE_DATA])
+
+    shades = Shades(pv_request)
+    shade_data = _async_map_data_by_id((await shades.get_resources())[SHADE_DATA])
+
+    async def async_update_data():
+        """Fetch data from shade endpoint."""
+        shades = hass.data[DOMAIN][entry.entry_id][PV_SHADES]
+        async with async_timeout.timeout(10):
+            shade_entries = await shades.get_resources()
+        if not shade_entries:
+            raise UpdateFailed(f"Failed to fetch new shade data.")
+        return _async_map_data_by_id(shade_entries[SHADE_DATA])
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="powerview hub",
+        update_method=async_update_data,
+        update_interval=timedelta(seconds=60),
+    )
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        PV_API: pv_request,
+        PV_ROOM_DATA: room_data,
+        PV_SCENE_DATA: scene_data,
+        PV_SHADES: shades,
+        PV_SHADE_DATA: shade_data,
+        COORDINATOR: coordinator,
+        DEVICE_INFO: device_info,
+    }
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
```

Comment:
One day we're going to come up with an elegant API so that we can guarantee that platforms have been set up once entry set up is done. Maybe an `entry.forward_setup(component)` that entry.async_setup checks once it is done calling `async_setup_entry`.